### PR TITLE
Add NamedTuple#merge(other : NamedTuple)

### PR DESCRIPTION
**Pull request by Papierkorb**
Hi,

Today in #crystal-lang a user required a possibility to merge two named tuples.
Their use-case was to call methods with named arguments from inside a RESTful API, while being able to easily set default values (or force other arguments to certain values).

**Sample code**

```crystal
# Read arguments from user
args = NamedTuple(user: String, password: String).from(rest_arg_hash)

# Force the `server` argument to a value
call_args = args.merge(server: "auth.my-service.com")

# And easily call the actual method
authenticate(**call_args)
```

Using this, one can easily do dynamic invocation on methods, possibly backed by further user-defined macro magic ;)

**Behaviour**

Takes two named tuples (The *other* can be kw-args style, or a proper named tuple), and merges keys and values from both into a new named tuple. If both tuples define a value for the same key, the value from the *other* is kept.

Cya

### CODE REVIEW

File: src/named_tuple.cr [hunk 1]
```diff
@@ -159,6 +159,28 @@ struct NamedTuple
...
+  # ```
+  def merge(other : NamedTuple)
+    merge(**other)
+  end
+
+  # ditto
+  def merge(**other : **U) forall U
+    {% begin %}
+    {
+      {% for k in T %} {% unless U.keys.includes?(k) %} "{{k}}": self[:"{{k}}"],{% end %} {% end %}
```
Comment (oprypin) [inline src/named_tuple.cr]: Please use `{{k.stringify}}` and not `"{{k}}"`. I think this is not safe.

**Comment by need47** [comment]
I have been long waiting on this. Currently I have my own merge method. Glad to see it built-in stdlib.

File: src/named_tuple.cr (hunk 1 above)
Comment (MakeNowJust) [inline src/named_tuple.cr]: Also, please use `{{k.symboloze}}` and not `:"{{k}}"`.

**Comment by Papierkorb** [comment]
Now using `stringify` and `symbolize`

**Comment by sdogruyol** [comment]
I also had a similar case. Having a `merge` for `NamedTuple` would definitely help.

**Comment by asterite** [comment]
Duplicate of #2634

I think I prefer the implementation of #2634

**Comment by Papierkorb** [comment]
#2634 blows up if both tuples have a value with the same key.

**Comment by emancu** [comment]
@Papierkorb that was the main idea, I was trying to _compose_ NamedTuples instead of merging them, so if you duplicate a key, it shouldn't compile.

In https://github.com/crystal-lang/crystal/pull/2634 you can see a [comment](https://github.com/crystal-lang/crystal/pull/2634#issuecomment-220980991) explaining why I  discarded the `merge` function, however, I found really useful any implementation, so happy to see this merged :)

**Comment by Papierkorb** [comment]
@emancu I totally see the worth in both methods, and I'd like to see both merged.

**Comment by straight-shoota** [comment]
Actually I think there are three reasonable methods of combining two `NamedTuple`s:

* `merge`: Creates a new `NamedTuple` type with the union of all keys. For the same key, values from one overwrite those of the other. Different types for the same key should either fail or combine them to a union type (?). The inversion should return the same `NamedTuple` type with possibly different values (iff keys are intersected). #4688
* `+`: Creates a new `NamedTuple` type with the union of all keys. Fails if `other` and `self` share a key. The operation is commutative. #4688
* `update`: Updates the values of `self` with those of `other` where `other`'s keys need to be a subset of `self`'s keys. Return type is the type of `self`. Inversion is only valid if the key sets are identical. The behavior  is similar to `merge` but it ensures that no additional keys are added and preserves the types from `self`.

For `update` there is no PR yet and I don't think it has been proposed before. I am not sure if the key and  type safety provided over `merge` adds much because mismatching types will either result in type errors elsewhere or shouldn't be to much of a concern.

**Comment by Sija** [comment]
@straight-shoota If I understand correctly there's no PR for `update` because `NamedTuple` is meant to be immutable.

**Comment by straight-shoota** [comment]
Of course it is. I'm sorry if this was misunderstanding, it would mean to create a *new* instance of the same `NamedTuple` type with updated values.

**Comment by Sija** [comment]
@Papierkorb Would be sad to have this PR buried, so 🏓

**Comment by asterite** [comment]
@Papierkorb Could you re-format the source files? It's only failing because of that.

File: src/named_tuple.cr [hunk 2]
```diff
@@ -159,6 +159,31 @@ struct NamedTuple
...
+    {
+      {% for k in T %} {% unless U.keys.includes?(k) %} {{k.stringify}}: self[{{k.symbolize}}],{% end %} {% end %}
+      {% for k in U %} {{k.stringify}}: other[{{k.symbolize}}], {% end %}
+    }
+    {% end %}
+  end
+
+  # Returns a hash value based on this name tuple's size, keys and values.
+  #
+  # See also: `Object#hash`.
```
Comment (Sija) [inline src/named_tuple.cr]: Redundant line.

**Comment by Papierkorb** [comment]
@asterite Ah shoot, I'm away from home this week-end.  Would it be okay if I do it on Monday?

**Comment by asterite** [comment]
@Papierkorb Of course! Take your time :-)

**Comment by Papierkorb** [comment]
Like this?

**Pull request merged**